File: Android - Native/discover_destination/app/src/main/java/com/example/discover_destination/screens/EditScreen.kt
package com.example.discover_destination.screens

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.OutlinedButton
import androidx.compose.material.OutlinedTextField
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import com.example.discover_destination.domain.Destination
import com.example.discover_destination.viewmodel.DestinationsViewModel

@Composable
fun EditScreen(
    viewModel: DestinationsViewModel = hiltViewModel(),
    user: String,
    city: String,
    onFinishedOperation: () -> Unit
){
    val destination = viewModel.getOne(city)
    val cityFieldValue = remember { mutableStateOf(TextFieldValue(destination.city)) }
    val countryFieldValue = remember { mutableStateOf(TextFieldValue(destination.country)) }
    val tourismFieldValue = remember { mutableStateOf(TextFieldValue(destination.tourism)) }
    val imageFieldValue = remember { mutableStateOf(TextFieldValue(destination.image)) }
    val descriptionFieldValue = remember { mutableStateOf(TextFieldValue(destination.description)) }
    val attractionsFieldValue = remember { mutableStateOf(TextFieldValue(destination.touristAttractions.reduce{acc: String, s: String -> acc + "\n" + s})) }
    var errorValue = remember { mutableStateOf(TextFieldValue())}

    Column(
        horizontalAlignment = Alignment.End,
        modifier = Modifier.padding(10.dp)
    ) {
        Text(
            text = "Edit Destination",
            fontSize = 35.sp
        )
        Column(
            horizontalAlignment = Alignment.Start
        ) {
            Text(
                text = "City",
                fontSize = 15.sp
            )
            OutlinedTextField(
                value = cityFieldValue.value,
                onValueChange = { cityFieldValue.value = it }
            )
            Text(
                text = "Country",
                fontSize = 15.sp
            )
            OutlinedTextField(
                value = countryFieldValue.value,
                onValueChange = { countryFieldValue.value = it }
            )
            Text(
                text = "Type of Tourism",
                fontSize = 15.sp
            )
            OutlinedTextField(
                value = tourismFieldValue.value,
                onValueChange = { tourismFieldValue.value = it }
            )
            Text(
                text = "Image URL",
                fontSize = 15.sp
            )
            OutlinedTextField(
                value = imageFieldValue.value,
                onValueChange = { imageFieldValue.value = it }
            )
            Text(
                text = "Description",
                fontSize = 15.sp
            )
            OutlinedTextField(
                value = descriptionFieldValue.value,
                onValueChange = { descriptionFieldValue.value = it }
            )
            Text(
                text = "List of Tourist Attractions:",
                fontSize = 15.sp
            )
            OutlinedTextField(
                value = attractionsFieldValue.value,
                onValueChange = { attractionsFieldValue.value = it }
            )
            OutlinedButton(onClick = {
                val destination = Destination(cityFieldValue.value.text, countryFieldValue.value.text,
                    tourismFieldValue.value.text, descriptionFieldValue.value.text,
                    imageFieldValue.value.text, attractionsFieldValue.value.text.split("\n"), user)
                if (validate(destination)){
                    viewModel.modify(destination)
                    onFinishedOperation()
                }
                else {
                    errorValue.value = TextFieldValue("Invalid data!")
                }
            },
                modifier = Modifier.size(400.dp, 55.dp)
            ) {
                Text(
                    text = "MODIFY DESTINATION",
                    fontSize = 20.sp,
                    color = Color.Black
                )
            }
            Text(
                text = errorValue.value.text,
                fontSize = 15.sp,
                color = Color.Red
            )
        }
    }
}

File: Flutter - Non-Native/discover_destination/lib/view/edit_screen.dart
import 'package:discover_destination/model/destination.dart';
import 'package:discover_destination/view/start_screen.dart';
import 'package:discover_destination/viewmodel/destinations_view_model.dart';
import 'package:flutter/material.dart';

class EditScreen extends StatefulWidget {
  DestinationsViewModel destinationsViewModel;
  Destination destination;

  EditScreen(this.destinationsViewModel, this.destination, {Key? key})
      : super(key: key);

  @override
  _EditState createState() => _EditState(destinationsViewModel, destination);
}

class _EditState extends State<EditScreen> {
  DestinationsViewModel destinationsViewModel;
  Destination destination;
  final cityTextFieldController = TextEditingController();
  final countryTextFieldController = TextEditingController();
  final tourismTextFieldController = TextEditingController();
  final imageTextFieldController = TextEditingController();
  final descriptionTextFieldController = TextEditingController();
  final attractionTextFieldController = TextEditingController();
  final attractionsTextFieldController = TextEditingController();

  _EditState(this.destinationsViewModel, this.destination);

  bool validate(Destination destination) {
    return destination.city.length > 0 &&
        destination.country.length > 0 &&
        destination.description.length > 0 &&
        destination.image.length > 0 &&
        destination.tourism.length > 0 &&
        !destination.touristAttractions.isEmpty;
  }

  @override
  Widget build(BuildContext context) {
    Destination newDestination;
    cityTextFieldController.text = destination.city;
    countryTextFieldController.text = destination.country;
    descriptionTextFieldController.text = destination.description;
    imageTextFieldController.text = destination.image;
    tourismTextFieldController.text = destination.tourism;
    attractionsTextFieldController.text = destination.touristAttractions
        .reduce((string, el) => string + ", " + el);
    List<String> attractions;

    return Scaffold(
        backgroundColor: Colors.white,
        body: SingleChildScrollView(
            child: Column(children: <Widget>[
          Padding(
            padding: const EdgeInsets.only(top: 60.0),
            child: Center(
              child: Container(
                  height: 70,
                  child: Text("Edit Destination",
                      textAlign: TextAlign.left,
                      style: TextStyle(fontSize: 45.0))),
            ),
          ),
          Padding(
              padding: EdgeInsets.symmetric(horizontal: 15),
              child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: <Widget>[
                    TextField(
                      decoration: InputDecoration(
                        border: OutlineInputBorder(),
                        labelText: 'City',
                      ),
                      controller: cityTextFieldController,
                    ),
                    SizedBox(
                      height: 10.0,
                    ),
                    TextField(
                      decoration: InputDecoration(
                        border: OutlineInputBorder(),
                        labelText: 'Country',
                      ),
                      controller: countryTextFieldController,
                    ),
                    SizedBox(
                      height: 10.0,
                    ),
                    TextField(
                      decoration: InputDecoration(
                        border: OutlineInputBorder(),
                        labelText: 'Type of Tourism',
                      ),
                      controller: tourismTextFieldController,
                    ),
                    SizedBox(
                      height: 10.0,
                    ),
                    TextField(
                      decoration: InputDecoration(
                        border: OutlineInputBorder(),
                        labelText: 'Image URL',
                      ),
                      controller: imageTextFieldController,
                    ),
                    SizedBox(
                      height: 10.0,
                    ),
                    TextField(
                      decoration: InputDecoration(
                        border: OutlineInputBorder(),
                        labelText: 'Description',
                      ),
                      controller: descriptionTextFieldController,
                    ),
                    SizedBox(
                      height: 10.0,
                    ),
                    Text("Tourist Attractions"),
                    TextField(
                      decoration: InputDecoration(border: OutlineInputBorder()),
                      controller: attractionsTextFieldController,
                    ),
                    SizedBox(
                      height: 10.0,
                    ),
                    Center(
                      child:
                    OutlinedButton(
                      onPressed: () => {
                        attractions =
                            attractionsTextFieldController.text.split(", "),
                        newDestination = Destination(
                            cityTextFieldController.text,
                            countryTextFieldController.text,
                            tourismTextFieldController.text,
                            descriptionTextFieldController.text,
                            imageTextFieldController.text,
                            attractions,
                            destination.user),
                        if (validate(newDestination))
                          {
                            destinationsViewModel.modify(newDestination),
                            Navigator.push(
                                context,
                                MaterialPageRoute(
                                    builder: (_) => StartScreen(
                                        destinationsViewModel,
                                        destination.user)))
                          }
                        else
                          {
                            showDialog(
                                context: context,
                                builder: (ctx) => AlertDialog(
                                      title: Text("Invalid data"),
                                      content:
                                          Text("All fields must be completed!"),
                                      actions: <Widget>[
                                        FlatButton(
                                          onPressed: () {
                                            Navigator.of(ctx).pop();
                                          },
                                          child: Text("OK"),
                                        ),
                                      ],
                                    ))
                          }
                      },
                      child: Text(
                        'SAVE DESTINATION',
                        style: TextStyle(color: Colors.black, fontSize: 25),
                      ),
                    ))
                  ]))
        ])));
  }
}
